### PR TITLE
Added Copy Link addon

### PR DESCRIPTION
**Pull request by Sly-Little-Fox**
**Resolves**

Nothing, just adds a new addon.

Resolves #
I added an addon to copy comment links. 

It is an addon.

I tested this addon on different versions of site and it works!

### CODE REVIEW

File: addons/copy-message-link/addon.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      "type": "boolean",
+      "default": false
+    },
+    {
+      "name": "Add button besides Report button",
+      "id": "report",
+      "type": "boolean",
+      "default": true
+    }
+  ],
```
Comment (TheColaber) [inline addons/copy-message-link/addon.json:21]: Make this one bool or make it a select type.

File: addons/copy-message-link/addon.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "Copy link button",
+  "description": "Adds \"Copy Link\" button to comments.",
```
Comment (TheColaber) [inline addons/copy-message-link/addon.json:3]: Maybe explain what exactly the link it copies is? Also, give yourself credit with the credits maifest key.

**Comment by RedGuy12** [comment]
Maybe turn the date into a link?
________________________________

File: addons/copy-message-link/addon.json (hunk 1 above)
Comment (Sly-Little-Fox) [reply addons/copy-message-link/addon.json:21]: Which one? They are already bools.

File: addons/copy-message-link/addon.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "Copy link button",
+  "description": "Adds \"Copy Link\" button to comments. Button copies link to comments, which you can use to show someone a comment. Works with profiles and studios.",
```
Comment (TheColaber) [inline addons/copy-message-link/addon.json:3]: Don't include
> Works with profiles and studios

File: addons/copy-message-link/addon.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  "credits": [
+    {
+      "name": "Sly-Little-Fox",
+      "link": "https://github.com/Sly-Little-Fox/"
+    }
+  ],
+  "userscripts": [{ 
+    "url": "userscript.js",
+    "matches": ["https://scratch.mit.edu/users/*", "https://scratch.mit.edu/studios/*"],
```
Comment (TheColaber) [inline addons/copy-message-link/addon.json:12]: Make sure to make this work on projects too!

File: addons/copy-message-link/addon.json (hunk 1 above)
Comment (TheColaber) [reply addons/copy-message-link/addon.json:27]: Yeah... make it either one bool or make it a select type.

File: addons/copy-message-link/userscript.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      newElem.textContent = 'Copy link';
+      newElem.setAttribute('onclick', `
+      let newElem2 = document.createElement('textarea');
+      newElem2.value = \`\${location.href.split('#')[0]}#\${elem.parentElement.parentElement.id}\`;
+      document.body.appendChild(newElem2);
+      newElem2.select();
+      document.execCommand('copy');
+      document.body.removeChild(newElem2);
+      delete newElem2;
+      `);
```
Comment (TheColaber) [inline addons/copy-message-link/userscript.js:43]: why not just `newElem.onclick = () => { /*code*/ }`

File: addons/copy-message-link/userscript.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      newElem.textContent = 'Copy Link';
+      newElem.setAttribute('onclick', `
+        let newElem2 = document.createElement('textarea');
+        newElem2.value = \`\${location.href.split('#')[0]}#\${this.parentElement.parentElement.parentElement.id}\`;
+        document.body.appendChild(newElem2);
+        newElem2.select();
+        document.execCommand('copy');
+        document.body.removeChild(newElem2);
+        delete newElem2;
+        `);
```
Comment (TheColaber) [inline addons/copy-message-link/userscript.js:24]: why not just newElem.onclick = () => { /*code*/ }

File: addons/copy-message-link/userscript.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
+export default async function ({
+  addon,
+  global,
+  console
+}) {
+  if (!addon.tab.clientVersion) return;
```
Comment (TheColaber) [inline addons/copy-message-link/userscript.js:6]: What's this for?

File: addons/copy-message-link/userscript.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  await addon.tab.waitForElement("div.comment", {
+    markAsSeen: true
+  });
+  if (addon.settings.get('reply')) {
+    document.querySelectorAll('div.comment').forEach((elem) => {
```
Comment (TheColaber) [inline addons/copy-message-link/userscript.js:11]: Something better to do here would to put the `await addon.tab.waitForElement` and set it to a variable, then you can just change that variable. You can read more about addon.tab.waitForElement in the wiki

**Review by TheColaber — CHANGES_REQUESTED** [review]

File: addons/copy-message-link/forum.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+export default async function ({
+  addon,
```
Comment (RedGuy12) [inline addons/copy-message-link/forum.js:2]: The date is already linked on the forms

Also where did studios go

File: addons/addons.json [hunk 10]
```diff
@@ -16,6 +16,7 @@
   "dark-www",
 
   "// Other addons:",
+  "copy-message-link",
```
Comment (WorldLanguages) [inline addons/addons.json:19]: Should be at the bottom, for consistency

File: addons/copy-message-link/addon.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "Copy link button",
+  "description": "Adds \"Copy Link\" button to comments. Button copies link to comments, which you can use to show someone a comment.",
+  "credits": [{
+    "name": "Sly-Little-Fox",
+    "link": "https://github.com/Sly-Little-Fox/"
+  }],
+  "userscripts": [{
+      "url": "profile.js",
+      "matches": ["https://scratch.mit.edu/users/*", "https://scratch.mit.edu/studios/*"],
```
Comment (WorldLanguages) [inline addons/copy-message-link/addon.json:10]: Matches more URLs than needed

File: addons/copy-message-link/addon.json [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    {
+      "url": "project.js",
+      "matches": ["https://scratch.mit.edu/projects/*"],
+      "runAtComplete": false
+    }
+  ],
+  "info": [{
+    "type": "notice",
+    "text": "Preferred button place will be ignored if it is not possible to place it button there.",
+    "id": "notice"
```
Comment (WorldLanguages) [inline addons/copy-message-link/addon.json:22]: Ideally this notice wouldn't be needed at all? Is it actually impossible to always place the button next to the report button?

File: addons/copy-message-link/forum.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
+// Dead userscript
```
Comment (WorldLanguages) [inline addons/copy-message-link/forum.js:1]: Should delete this file

File: addons/copy-message-link/forum.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  addon,
+  global,
+  console
+}) {
+  await addon.tab.waitForElement("div.blockpost", {
+    markAsSeen: true
+  });
+  document.querySelectorAll('div.blockpost').forEach((elem) => {
+    let newElem = document.createElement('a');
+    newElem.setAttribute('nohref', '');
```
Comment (WorldLanguages) [inline addons/copy-message-link/forum.js:13]: Does nohref mean anything in newer browsers?

File: addons/copy-message-link/profile.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    markAsSeen: true
+  });
+  if (addon.settings.get('preferred') === 'reply') {
+    document.querySelectorAll('div.comment').forEach((elem) => {
+      let newElem = document.createElement('a');
+      newElem.className = 'reply';
+      newElem.textContent = 'Copy Link';
+      newElem.onclick = (e) => {
+        let newElem2 = document.createElement('textarea');
+        newElem2.value = `${location.href.split('#')[0]}#${e.target.parentElement.parentElement.parentElement.id}`;
```
Comment (WorldLanguages) [inline addons/copy-message-link/profile.js:16]: Would be cool if it used proper casing for the username - URL might say `world_languaGES` but it should instead be `World_Languages`

File: addons/copy-message-link/profile.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+      newElem.setAttribute('nohref', '');
+      newElem.className = 'actions report';
+      newElem.style = 'cursor: pointer;';
+      newElem.textContent = 'Copy link';
+      newElem.onclick = (e) => {
+        let newElem2 = document.createElement('textarea');
+        newElem2.value = `${location.href.split('#')[0]}#${e.target.parentElement.parentElement.id}`;
+        document.body.appendChild(newElem2);
+        newElem2.select();
+        document.execCommand('copy');
```
Comment (WorldLanguages) [inline addons/copy-message-link/profile.js:37]: `navigator.clipboard.writeText()` is a better alternative that works on browsers we support

File: addons/copy-message-link/profile.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+      newElem.setAttribute('nohref', 'nohref');
+      document.querySelector(`div#${elem.id} > div.info > div:nth-child(3)`).appendChild(newElem);
+    });
+  } else if (addon.settings.get('preferred') === 'report') {
+    document.querySelectorAll('div.comment').forEach((elem) => {
+      let newElem = document.createElement('span');
+      newElem.setAttribute('nohref', '');
+      newElem.className = 'actions report';
+      newElem.style = 'cursor: pointer;';
+      newElem.textContent = 'Copy link';
```
Comment (WorldLanguages) [inline addons/copy-message-link/profile.js:31]: Let's localize this string - also, is it "Copy link" or "Copy Link"? Varies between profile.js and project.js

Comment (mxmou) [reply addons/copy-message-link/profile.js:31]: > also, is it "Copy link" or "Copy Link"? Varies between profile.js and project.js

That's probably for consistency with other strings on those pages.

**Pull request merged**